File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/groupsv2/ProtobufTestUtils.java
package org.whispersystems.signalservice.api.groupsv2;

import com.squareup.wire.Message;
import com.squareup.wire.WireField;

import java.util.stream.Stream;

final class ProtobufTestUtils {

  /**
   * Finds the largest declared field number in the supplied protobuf class.
   */
  static int getMaxDeclaredFieldNumber(Class<? extends Message<?, ?>> protobufClass) {
    return Stream.of(protobufClass.getFields())
                 .map(f -> f.getAnnotationsByType(WireField.class))
                 .filter(a -> a.length == 1)
                 .map(a -> a[0].tag())
                 .max(Integer::compareTo)
                 .orElse(0);
  }
}

File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/groupsv2/ProtobufTestUtils.kt
package org.whispersystems.signalservice.api.groupsv2

import com.squareup.wire.Message
import com.squareup.wire.WireField

internal object ProtobufTestUtils {
  /**
   * Finds the largest declared field number in the supplied protobuf class.
   */
  fun getMaxDeclaredFieldNumber(protobufClass: Class<out Message<*, *>?>): Int {
    return protobufClass.fields
      .map { field -> field.getAnnotationsByType(WireField::class.java) }
      .filter { array -> array.size == 1 }
      .maxOfOrNull { array -> array.first().tag }
      ?: 0
  }
}
